firewood/src/storage/linear/proposed.rs

```diff
@@ -7,39 +7,55 @@ use std::io::{Error, Read};
 use std::marker::PhantomData;
 use std::sync::Arc;
 
-use super::{LinearStore, ReadLinearStore, WriteLinearStore};
+use super::layered::{Layer, LayeredReader};
+use super::ImmutableLinearStore;
 
 /// [Proposed] is a [LinearStore] state that contains a copy of the old and new data.
 /// The P type parameter indicates the state of the linear store for it's parent,
 /// which could be a another [Proposed] or is [FileBacked](super::filebacked::FileBacked)
 /// The M type parameter indicates the mutability of the proposal, either read-write or readonly
+/// TODO update comment
 #[derive(Debug)]
-pub(crate) struct Proposed<P: ReadLinearStore, M> {
-    pub(crate) new: BTreeMap<u64, Box<[u8]>>,
-    old: BTreeMap<u64, Box<[u8]>>,
-    pub(crate) parent: Arc<LinearStore<P>>,
-    phantom: PhantomData<M>,
+pub(crate) struct Proposed<T> {
+    new: BTreeMap<u64, Box<[u8]>>,
+    pub(super) old: BTreeMap<u64, Box<[u8]>>,
+    pub(super) parent: Arc<ImmutableLinearStore>,
+    state: PhantomData<T>,
 }
 
-impl<P: ReadLinearStore, M> Proposed<P, M> {
-    pub(crate) fn new(parent: Arc<LinearStore<P>>) -> Self {
+impl From<Proposed<Mutable>> for Proposed<Immutable> {
```

Comment:
This is a nit, but I think it would be more explicit using a method named something like `freeze`. 

The call site goes from this
```rust
let proposal: Proposal<Immutable> = proposal.into();
// or 
let proposal = Proposal::<Immutable>::from(proposal);
```
to
```rust
let proposal = proposal.freeze();
```
Let Rust-Analyzer give you the type hints from there.